File: CSharpFundamentals/CSharpAdvanced/StacksAndQueues_Ex/BalancedParenthesis.cs
using System;
using System.Collections.Generic;

namespace BalancedParenthesis
{
    class BalancedParenthesis
    {
        private static Stack<char> stack = new Stack<char>();
        private static char[] parentheses;
        static void Main(string[] args)
        {
            parentheses = Console.ReadLine().ToCharArray();
            if (AlgoBalancedParentheses())
            {
                Console.WriteLine("YES");
            }
            else
            {
                Console.WriteLine("NO");
            }
        }

        private static bool AlgoBalancedParentheses()
        {
            foreach (char symbol in parentheses)
            {
                if (isOpen(symbol))
                {
                    stack.Push(symbol);
                }
                else
                {
                    if (stack.Count == 0)
                    {
                        return false;
                    }
                    if (!checkMatch(stack.Pop(), symbol))
                    {
                        return false;
                    }
                }
            }

            return stack.Count == 0;
        }

        private static bool checkMatch(char pop, char symbol)
        {
            return (pop == '{' && symbol == '}')
                || (pop == '[' && symbol == ']')
                || (pop == '(' && symbol == ')');
        }

        private static bool isOpen(char c)
        {
            return c == '{' || c == '[' || c == '(';
        }
    }
}

File: JavaFundamentals/JavaAdvanced/Stacks-and-queues-ex/src/BalancedParenthesis.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;

public class BalancedParenthesis {
    private static ArrayDeque<Character> stack = new ArrayDeque<>();
    private static char[] parentheses;

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        parentheses = reader.readLine().toCharArray();
        if (algoBalancedParenteses()){
            System.out.println("YES");
        } else {
            System.out.println("NO");
        }
    }

    private static boolean algoBalancedParenteses(){
        for (char symbol : parentheses) {
            if (isOpen(symbol)) {
                stack.push(symbol);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                if (!checkMatch(stack.pop(), symbol)) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }

    private static boolean checkMatch(Character pop, char symbol) {
        return (pop == '{' && symbol == '}')
                || (pop == '[' && symbol == ']')
                || (pop == '(' && symbol == ')');
    }

    private static boolean isOpen(char c) {
        return c == '{' || c == '[' || c == '(';
    }
}
